plugins/pulp_rpm/plugins/importers/yum/upload.py

```diff
@@ -126,58 +145,49 @@ def _handle_erratum(repo, type_id, unit_key, metadata, file_path, conduit, confi
     steps are to save the metadata and optionally link the erratum to RPMs
     in the repository.
 
-    :type  repo: pulp.plugins.model.Repository
+    :param repo: The repository to import the package into
+    :type  repo: pulp.server.db.model.Repository
+
+    :param type_id: The type_id of the package being uploaded
     :type  type_id: str
+
+    :param unit_key: A dictionary of fields to overwrite introspected field values
     :type  unit_key: dict
+
+    :param metadata: A dictionary of fields to overwrite introspected field values, or None
     :type  metadata: dict or None
+
+    :param file_path: The path to the uploaded package
     :type  file_path: str
+
+    :param conduit: provides access to relevant Pulp functionality
     :type  conduit: pulp.plugins.conduits.upload.UploadConduit
+
+    :param config: plugin configuration for the repository
     :type  config: pulp.plugins.config.PluginCallConfiguration
     """
+    model_class = plugin_api.get_unit_model_by_id(type_id)
+    update_fields_inbound(model_class, unit_key or {})
+    update_fields_inbound(model_class, metadata or {})
 
-    # Validate the user specified data by instantiating the model
-    try:
-        model_class = models.TYPE_MAP[type_id]
-        model = model_class(metadata=metadata, **unit_key)
-    except TypeError:
-        raise ModelInstantiationError()
+    unit_data = {}
+    unit_data.update(metadata or {})
+    unit_data.update(unit_key or {})
 
-    unit = conduit.init_unit(model.TYPE, model.unit_key, model.metadata, None)
+    unit = model_class(**unit_data)
 
-    # this save must happen before the link is created, because the link logic
-    # requires the unit to have an "id".
-    saved_unit = conduit.save_unit(unit)
+    unit.save()
 
     if not config.get_boolean(CONFIG_SKIP_ERRATUM_LINK):
-        _link_errata_to_rpms(conduit, model, saved_unit)
-
-
-def _link_errata_to_rpms(conduit, errata_model, errata_unit):
-    """
-    Creates links in the Pulp data model between an erratum and its RPMs.
-
-    :param conduit: provides access to relevant Pulp functionality
-    :type  conduit: pulp.plugins.conduits.unit_add.UnitAddConduit
-    :param errata_model:    model object representing an errata
-    :type  errata_model:    pulp_rpm.plugins.db.models.Errata
-    :param errata_unit:     unit object representing an errata
-    :type  errata_unit:     pulp.plugins.model.Unit
-    """
-    fields = list(models.RPM.UNIT_KEY_NAMES)
-    fields.append('_storage_path')
-    filters = {'$or': errata_model.rpm_search_dicts}
-    for model_type in (models.RPM.TYPE, models.SRPM.TYPE):
-        criteria = UnitAssociationCriteria(type_ids=[model_type], unit_fields=fields,
-                                           unit_filters=filters)
-        for unit in conduit.get_units(criteria):
-            conduit.link_unit(errata_unit, unit, bidirectional=True)
+        for model_type in [models.RPM.TYPE, models.SRPM.TYPE]:
```

Comment:
I adjusted it so that it won't raise an exception, but it's still not finished. This is the exact line that my upload work left off at. Completing the TODO below it will be picked up as part of a post merge redmine task.